File: sample/cpp/sample_img_hand_keypoint_gesture.cpp
#include "tdl_model_factory.hpp"
std::vector<cv::Scalar> color = {
    cv::Scalar(51, 153, 255), cv::Scalar(0, 153, 76), cv::Scalar(255, 215, 0),
    cv::Scalar(255, 128, 0), cv::Scalar(0, 255, 0)};

void visualize_keypoints_detection(std::shared_ptr<BaseImage> image,
                                   std::shared_ptr<ModelLandmarksInfo> obj_meta,
                                   float score, const std::string &save_path) {
  cv::Mat mat;
  bool is_rgb;
  int32_t ret = ImageFactory::convertToMat(image, mat, is_rgb);
  if (ret != 0) {
    std::cout << "Failed to convert to mat" << std::endl;
    return;
  }

  if (is_rgb) {
    std::cout << "convert to bgr" << std::endl;
    cv::cvtColor(mat, mat, cv::COLOR_RGB2BGR);
  }
  uint32_t image_width = obj_meta->image_width;
  uint32_t image_height = obj_meta->image_height;

  for (uint32_t j = 0; j < 21; j++) {
    int x = static_cast<int>(obj_meta->landmarks_x[j] * image_width);
    int y = static_cast<int>(obj_meta->landmarks_y[j] * image_height);
    cv::circle(mat, cv::Point(x, y), 2, cv::Scalar(0, 0, 255), -1);
  }

  cv::imwrite(save_path, mat);
}

std::vector<std::shared_ptr<ModelOutputInfo>> extract_crop_hand_landmark(
    std::shared_ptr<BaseModel> model_hk,
    std::vector<std::shared_ptr<BaseImage>> images,
    std::vector<std::shared_ptr<BaseImage>> &hand_crops,
    std::vector<std::shared_ptr<ModelOutputInfo>> &hand_metas) {
  std::shared_ptr<BasePreprocessor> preprocessor = model_hk->getPreprocessor();

  std::vector<std::shared_ptr<ModelOutputInfo>> out_datas;
  char sz_img_name[128];
  for (size_t b = 0; b < images.size(); b++) {
    std::shared_ptr<ModelBoxInfo> hand_meta =
        std::static_pointer_cast<ModelBoxInfo>(hand_metas[b]);

    for (size_t i = 0; i < hand_meta->bboxes.size(); i++) {
      int x1 = hand_meta->bboxes[i].x1;
      int y1 = hand_meta->bboxes[i].y1;
      int x2 = hand_meta->bboxes[i].x2;
      int y2 = hand_meta->bboxes[i].y2;

      int width = x2 - x1;
      int height = y2 - y1;

      float expansion_factor = 1.25f;
      int new_width = static_cast<int>(width * expansion_factor);
      int new_height = static_cast<int>(height * expansion_factor);

      int crop_x1 = x1 - (new_width - width) / 2;
      int crop_y1 = y1 - (new_height - height) / 2;
      int crop_x2 = crop_x1 + new_width;
      int crop_y2 = crop_y1 + new_height;

      std::shared_ptr<BaseImage> hand_crop = preprocessor->crop(
          images[b], crop_x1, crop_y1, crop_x2 - crop_x1, crop_y2 - crop_y1);

      sprintf(sz_img_name, "hand_crop_%ld.jpg", i);
      ImageFactory::writeImage(sz_img_name, hand_crop);
      hand_crops.push_back(hand_crop);
    }
  }
  model_hk->inference(hand_crops, out_datas);

  return out_datas;
}

std::vector<std::shared_ptr<ModelOutputInfo>> hand_keypoint_classification(
    std::shared_ptr<BaseModel> model_hc,
    std::vector<std::shared_ptr<ModelOutputInfo>> &out_hk) {
  std::vector<std::shared_ptr<BaseImage>> input_datas;
  std::vector<std::shared_ptr<ModelOutputInfo>> out_datas;
  for (size_t i = 0; i < out_hk.size(); i++) {
    std::shared_ptr<ModelLandmarksInfo> obj_meta =
        std::static_pointer_cast<ModelLandmarksInfo>(out_hk[i]);

    std::vector<float> keypoints;

    for (uint32_t k = 0; k < 21; k++) {
      keypoints.push_back(obj_meta->landmarks_x[k]);
      keypoints.push_back(obj_meta->landmarks_y[k]);
    }

    std::shared_ptr<BaseImage> bin_data = ImageFactory::createImage(
        42, 1, ImageFormat::GRAY, TDLDataType::FP32, true);

    float *data_buffer =
        reinterpret_cast<float *>(bin_data->getVirtualAddress()[0]);

    memcpy(data_buffer, &keypoints[0], 42 * sizeof(float));
    input_datas.push_back(bin_data);
  }

  model_hc->inference(input_datas, out_datas);
  return out_datas;
}

int main(int argc, char **argv) {
  if (argc != 3) {
    printf("Usage: %s <model_dir> <image_path> \n", argv[0]);
    return -1;
  }
  std::string model_dir = argv[1];
  std::string image_path = argv[2];

  std::shared_ptr<BaseImage> image1 = ImageFactory::readImage(image_path);
  if (!image1) {
    printf("Failed to create image1\n");
    return -1;
  }

  TDLModelFactory &model_factory = TDLModelFactory::getInstance();
  model_factory.loadModelConfig();
  model_factory.setModelDir(model_dir);

  std::shared_ptr<BaseModel> model_hd =
      model_factory.getModel(ModelType::YOLOV8N_DET_HAND);
  if (!model_hd) {
    printf("Failed to create model_hd\n");
    return -1;
  }

  std::shared_ptr<BaseModel> model_hk =
      model_factory.getModel(ModelType::KEYPOINT_HAND);
  if (!model_hk) {
    printf("Failed to create model_hk\n");
    return -1;
  }

  std::shared_ptr<BaseModel> model_hc =
      model_factory.getModel(ModelType::CLS_KEYPOINT_HAND_GESTURE);
  if (!model_hc) {
    printf("Failed to create model_hr\n");
    return -1;
  }

  std::vector<std::shared_ptr<ModelOutputInfo>> out_hd;
  std::vector<std::shared_ptr<BaseImage>> input_images = {image1};
  std::vector<std::shared_ptr<BaseImage>> hand_crops;
  model_hd->inference(input_images, out_hd);
  std::vector<std::shared_ptr<ModelOutputInfo>> out_hk =
      extract_crop_hand_landmark(model_hk, input_images, hand_crops, out_hd);

  std::vector<std::shared_ptr<ModelOutputInfo>> out_hc =
      hand_keypoint_classification(model_hc, out_hk);

  for (size_t i = 0; i < out_hk.size(); i++) {
    std::shared_ptr<ModelLandmarksInfo> obj_meta =
        std::static_pointer_cast<ModelLandmarksInfo>(out_hk[i]);

    std::shared_ptr<ModelClassificationInfo> cls_meta =
        std::static_pointer_cast<ModelClassificationInfo>(out_hc[i]);

    for (int k = 0; k < 21; k++) {
      printf("%d: %f %f\n", k, obj_meta->landmarks_x[k] * obj_meta->image_width,
             obj_meta->landmarks_y[k] * obj_meta->image_height);
    }
    printf("hand[%ld]: label: %d, score: %.2f\n", i, cls_meta->topk_class_ids[0],
           cls_meta->topk_scores[0]);

    char img_name[128];
    sprintf(img_name, "hand_keypoints_%ld_label_%d.jpg", i,
            cls_meta->topk_class_ids[0]);
    visualize_keypoints_detection(hand_crops[i], obj_meta, 0.5,
                                  std::string(img_name));
  }

  return 0;
}

File: sample/python/sample_img_hand_keypoint_gesture.py
import sys
import os
import cv2
from tdl import nn, image
import numpy as np

def crop_and_expand(img, bbox, expansion=1.25):
    x1, y1, x2, y2 = bbox
    width = x2 - x1
    height = y2 - y1
    new_width = int(width * expansion)
    new_height = int(height * expansion)
    crop_x1 = int(max(x1 - (new_width - width) // 2, 0))
    crop_y1 = int(max(y1 - (new_height - height) // 2, 0))
    crop_x2 = int(min(crop_x1 + new_width, img.shape[1]))
    crop_y2 = int(min(crop_y1 + new_height, img.shape[0]))
    crop_img = img[crop_y1:crop_y2, crop_x1:crop_x2].copy()
    return crop_img, (crop_x1, crop_y1)


def visualize_keypoints(img, landmarks, save_path):
    for x, y in landmarks:
        cv2.circle(img, (int(x), int(y)), 2, (0, 0, 255), -1)
    cv2.imwrite(save_path, img)
    print(f"保存图像到: {save_path}")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python3 sample_img_hand_keypoint_gesture.py <model_dir> <image_path>")
        sys.exit(1)

    model_dir = sys.argv[1]
    image_path = sys.argv[2]

    model_hd = nn.get_model_from_dir(nn.ModelType.YOLOV8N_DET_HAND, model_dir)
    model_hk = nn.get_model_from_dir(nn.ModelType.KEYPOINT_HAND, model_dir)
    model_hc = nn.get_model_from_dir(nn.ModelType.CLS_KEYPOINT_HAND_GESTURE, model_dir)

    if not os.path.exists(image_path):
        print(f"图像文件不存在: {image_path}")
        sys.exit(1)
    
    img = image.read(image_path)
    img_cv = cv2.imread(image_path)

    out_hd = model_hd.inference(img)
    if len(out_hd) == 0:
        print("No hand detected")
        sys.exit(0)

    hand_crops = []
    crop_offsets = []
    for i, box in enumerate(out_hd):
        x1, y1, x2, y2 = int(box["x1"]), int(box["y1"]), int(box["x2"]), int(box["y2"])
        crop_img, (off_x, off_y) = crop_and_expand(img_cv, (x1, y1, x2, y2), expansion=1.25)
        crop_offsets.append((off_x, off_y))
        crop_img_path = f"hand_crop_{i}.jpg"
        cv2.imwrite(crop_img_path, crop_img)
        hand_crops.append(image.read(crop_img_path))

    out_hk = []
    for crop in hand_crops:
        out_hk.extend(model_hk.inference(crop))


    out_hc = []
    for hk in out_hk:
        keypoints = []
        for pt in hk["landmarks"]:
            keypoints.extend(pt)
        # 关键点输入为 42x1x1 float32 单通道
        bin_data = image.Image.from_numpy(
            np.array(keypoints, dtype="float32").reshape(42, 1, 1),
            image.ImageFormat.GRAY
        )
        out_hc.extend(model_hc.inference(bin_data))

    for i, (hk, hc) in enumerate(zip(out_hk, out_hc)):
        landmarks = []
        for idx, pt in enumerate(hk["landmarks"]):
            x = pt[0] * hand_crops[i].get_size()[0] + crop_offsets[i][0]
            y = pt[1] * hand_crops[i].get_size()[1] + crop_offsets[i][1]
            landmarks.append((x, y))
            print(f"{idx}: {x:.2f} {y:.2f}")
        print(f"hand[{i}]: label: {hc['class_id']}, score: {hc['score']:.2f}")
        save_path = f"hand_keypoints_{i}_label_{hc['class_id']}.jpg"
        visualize_keypoints(img_cv.copy(), landmarks, save_path)

# input: python3 sample_img_hand_keypoint_gesture.py <model_dir> <image_path>
# output: hand[i]: label: <class_id>, score: <score>